File: tests/src/test/javascript/entityauth/PresharedAuthenticationFactoryTest.js
/**
 * Copyright (c) 2012-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Pre-shared keys entity authentication factory unit tests.
 * 
 * @author Wesley Miaw <wmiaw@netflix.com>
 */
describe("PresharedAuthenticationFactory", function() {
    var MslEncoderFormat = require('msl-core/io/MslEncoderFormat.js');
    var EntityAuthenticationScheme = require('msl-core/entityauth/EntityAuthenticationScheme.js');
    var PresharedAuthenticationFactory = require('msl-core/entityauth/PresharedAuthenticationFactory.js');
    var PresharedAuthenticationData = require('msl-core/entityauth/PresharedAuthenticationData.js');
    var MslEncoderUtils = require('msl-core/io/MslEncoderUtils.js');
    var MslEncodingException = require('msl-core/MslEncodingException.js');
    var MslEntityAuthException = require('msl-core/MslEntityAuthException.js');
    var MslError = require('msl-core/MslError.js');

    var MslTestConstants = require('msl-tests/MslTestConstants.js');
    var MockAuthenticationUtils = require('msl-tests/util/MockAuthenticationUtils.js');
    var MockMslContext = require('msl-tests/util/MockMslContext.js');
    var MockKeySetStore = require('msl-tests/entityauth/MockKeySetStore.js');
    var MockPresharedAuthenticationFactory = require('msl-tests/entityauth/MockPresharedAuthenticationFactory.js');
    var MslTestUtils = require('msl-tests/util/MslTestUtils.js');
    
    /** MSL encoder format. */
    var ENCODER_FORMAT = MslEncoderFormat.JSON;
    
    /** Key entity identity. */
    var KEY_IDENTITY = "identity";

    /** Authentication utilities. */
    var authutils = new MockAuthenticationUtils();
    /** Entity authentication factory. */
    var factory;
    /** MSL context. */
    var ctx;
    /** MSL encoder factory. */
    var encoder;
    
    var initialized = false;
    beforeEach(function() {
        if (!initialized) {
            runs(function() {
                MockMslContext.create(EntityAuthenticationScheme.PSK, false, {
                    result: function(c) { ctx = c; },
                    error: function(e) { expect(function() { throw e; }).not.toThrow(); }
                });
            });
            waitsFor(function() { return ctx; }, "ctx", MslTestConstants.TIMEOUT_CTX);
            runs(function() {
                encoder = ctx.getMslEncoderFactory();
                var store = new MockKeySetStore();
                store.addKeys(MockPresharedAuthenticationFactory.PSK_ESN, MockPresharedAuthenticationFactory.KPE, MockPresharedAuthenticationFactory.KPH, MockPresharedAuthenticationFactory.KPW);
                factory = new PresharedAuthenticationFactory(store, authutils);
                ctx.addEntityAuthenticationFactory(factory);
                initialized = true;
            });
            waitsFor(function() { return initialized; }, "static initialization", MslTestConstants.TIMEOUT);
        }
    });
    
    afterEach(function() {
        authutils.reset();
    });
    
    it("createData", function () {
        var data = new PresharedAuthenticationData(MockPresharedAuthenticationFactory.PSK_ESN);
        var entityAuthMo;
        runs(function() {
            data.getAuthData(encoder, ENCODER_FORMAT, {
                result: function(x) { entityAuthMo = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); }
            });
        });
        waitsFor(function() { return entityAuthMo; }, "entityAuthMo", MslTestConstants.TIMEOUT);
        
        var authdata;
        runs(function() {
            factory.createData(ctx, entityAuthMo, {
                result: function(x) { authdata = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); }
            });
        });
        waitsFor(function() { return authdata; }, "authdata", MslTestConstants.TIMEOUT);

        var dataMo, authdataMo;
        runs(function() {
            expect(authdata).not.toBeNull();
            expect(authdata instanceof PresharedAuthenticationData).toBeTruthy();
            
            MslTestUtils.toMslObject(encoder, data, {
                result: function(x) { dataMo = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); }
            });
            MslTestUtils.toMslObject(encoder, authdata, {
                result: function(x) { authdataMo = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); }
            });
        });
        waitsFor(function() { return dataMo && authdataMo; }, "dataMo && authdataMo", MslTestConstants.TIMEOUT);

        runs(function() {
            expect(MslEncoderUtils.equalObjects(dataMo, authdataMo)).toBeTruthy();
        });
    });
    
    it("encode exception", function() {
        var entityAuthMo;
        runs(function() {
	        var data = new PresharedAuthenticationData(MockPresharedAuthenticationFactory.PSK_ESN);
	        data.getAuthData(encoder, ENCODER_FORMAT, {
	            result: function(x) { entityAuthMo = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); }
	        });
        });
        waitsFor(function() { return entityAuthMo; }, "entityAuthMo", MslTestConstants.TIMEOUT);
        
        var exception;
        runs(function() {
	        entityAuthMo.remove(KEY_IDENTITY);
	        factory.createData(ctx, entityAuthMo, {
                result: function() {},
                error: function(e) { exception = e; },
	        });
        });
    	waitsFor(function() { return exception; }, "exception", MslTestConstants.TIMEOUT);
    	
    	runs(function() {
    	    var f = function() { throw exception; };
    	    expect(f).toThrow(new MslEncodingException(MslError.MSL_PARSE_ERROR));
    	});
    });
    
    it("crypto context", function() {
        var data = new PresharedAuthenticationData(MockPresharedAuthenticationFactory.PSK_ESN);
        var cryptoContext = factory.getCryptoContext(ctx, data);
        expect(cryptoContext).not.toBeNull();
    });
    
    it("unknown ESN", function() {
    	var f = function() {
	        var data = new PresharedAuthenticationData(MockPresharedAuthenticationFactory.PSK_ESN2);
	        factory.getCryptoContext(ctx, data);
    	};
    	expect(f).toThrow(new MslEntityAuthException(MslError.ENTITY_NOT_FOUND));
    });
    
    it("revoked", function() {
        var f = function() {
            authutils.revokeEntity(MockPresharedAuthenticationFactory.PSK_ESN);
            var data = new PresharedAuthenticationData(MockPresharedAuthenticationFactory.PSK_ESN);
            factory.getCryptoContext(ctx, data);
        };
        expect(f).toThrow(new MslEntityAuthException(MslError.ENTITY_REVOKED));
    });
});

File: tests/src/test/cpp/entityauth/PresharedAuthenticationFactoryTest.cpp
/**
 * Copyright (c) 2016-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <entityauth/MockKeySetStore.h>
#include <entityauth/PresharedAuthenticationData.h>
#include <gtest/gtest.h>
#include <gmock/gmock.h>
#include <entityauth/PresharedAuthenticationFactory.h>
#include <io/MslEncoderFactory.h>
#include <io/MslEncoderFormat.h>
#include <Macros.h>
#include <MslEncodingException.h>
#include <MslEntityAuthException.h>
#include <memory>
#include <string>

#include <entityauth/MockPresharedAuthenticationFactory.h>
#include <util/MockAuthenticationUtils.h>
#include <util/MockMslContext.h>
#include <util/MslTestUtils.h>

using namespace std;
using namespace testing;
using namespace netflix::msl;
using namespace netflix::msl::crypto;
using namespace netflix::msl::entityauth;
using namespace netflix::msl::io;
using namespace netflix::msl::util;

namespace netflix {
namespace msl {
namespace util {

namespace {

/** Key entity identity. */
const string KEY_IDENTITY = "identity";

} // namespace anonymous

/**
 * Pre-shared keys entity authentication factory unit tests.
 */
class PresharedAuthenticationFactoryTest : public ::testing::Test
{
public:
	PresharedAuthenticationFactoryTest()
    : ENCODER_FORMAT(MslEncoderFormat::JSON)
    , ctx(make_shared<MockMslContext>(EntityAuthenticationScheme::PSK, false))
    , encoder(ctx->getMslEncoderFactory())
    , authutils(make_shared<MockAuthenticationUtils>())
    {
	    shared_ptr<MockKeySetStore> store = make_shared<MockKeySetStore>();
        store->addKeys(MockPresharedAuthenticationFactory::PSK_ESN,
                MockPresharedAuthenticationFactory::KPE,
                MockPresharedAuthenticationFactory::KPH,
                MockPresharedAuthenticationFactory::KPW);
        factory = make_shared<PresharedAuthenticationFactory>(store, authutils);
        ctx->addEntityAuthenticationFactory(factory);
    }
	~PresharedAuthenticationFactoryTest() { authutils->reset(); }
protected:
    /** MSL encoder format. */
    const MslEncoderFormat ENCODER_FORMAT;
    /** MSL context. */
    shared_ptr<MockMslContext> ctx;
    /** MSL encoder factory-> */
    shared_ptr<MslEncoderFactory> encoder;
    /** Authentication utilities. */
    shared_ptr<MockAuthenticationUtils> authutils;
    /** Entity authentication factory-> */
    shared_ptr<EntityAuthenticationFactory> factory;
};

TEST_F(PresharedAuthenticationFactoryTest, createData)
{
    shared_ptr<PresharedAuthenticationData> data = make_shared<PresharedAuthenticationData>(MockPresharedAuthenticationFactory::PSK_ESN);
    shared_ptr<MslObject> entityAuthMo = data->getAuthData(encoder, ENCODER_FORMAT);

    shared_ptr<EntityAuthenticationData> authdata = factory->createData(ctx, entityAuthMo);
    EXPECT_TRUE(authdata);
    EXPECT_TRUE(instanceof<PresharedAuthenticationData>(authdata.get()));

    shared_ptr<MslObject> dataMo = MslTestUtils::toMslObject(encoder, data);
    shared_ptr<MslObject> authdataMo = MslTestUtils::toMslObject(encoder, authdata);
    EXPECT_EQ(*dataMo, *authdataMo);
}

TEST_F(PresharedAuthenticationFactoryTest, encodeException)
{
//    thrown.expect(MslEncodingException.class);
//    thrown.expectMslError(MslError.MSL_PARSE_ERROR);

    shared_ptr<PresharedAuthenticationData> data = make_shared<PresharedAuthenticationData>(MockPresharedAuthenticationFactory::PSK_ESN);
    shared_ptr<MslObject> entityAuthMo = data->getAuthData(encoder, ENCODER_FORMAT);
    entityAuthMo->remove(KEY_IDENTITY);
    try {
        factory->createData(ctx, entityAuthMo);
        ADD_FAILURE() << "Should have thrown.";
    } catch (const MslEncodingException& e) {
        EXPECT_EQ(MslError::MSL_PARSE_ERROR, e.getError());
    }
}

TEST_F(PresharedAuthenticationFactoryTest, cryptoContext)
{
    shared_ptr<PresharedAuthenticationData> data = make_shared<PresharedAuthenticationData>(MockPresharedAuthenticationFactory::PSK_ESN);
    shared_ptr<ICryptoContext> cryptoContext = factory->getCryptoContext(ctx, data);
    EXPECT_TRUE(cryptoContext);
}

TEST_F(PresharedAuthenticationFactoryTest, unknownEsn)
{
//    thrown.expect(MslEntityAuthException.class);
//    thrown.expectMslError(MslError.ENTITY_NOT_FOUND);

    shared_ptr<PresharedAuthenticationData> data = make_shared<PresharedAuthenticationData>(MockPresharedAuthenticationFactory::PSK_ESN2);
    try {
        factory->getCryptoContext(ctx, data);
        ADD_FAILURE() << "Should have thrown.";
    } catch (const MslEntityAuthException& e) {
        EXPECT_EQ(MslError::ENTITY_NOT_FOUND, e.getError());
    }
}

TEST_F(PresharedAuthenticationFactoryTest, revoked)
{
//    thrown.expect(MslEntityAuthException.class);
//    thrown.expectMslError(MslError.ENTITY_REVOKED);

    authutils->revokeEntity(MockPresharedAuthenticationFactory::PSK_ESN);
    shared_ptr<PresharedAuthenticationData> data = make_shared<PresharedAuthenticationData>(MockPresharedAuthenticationFactory::PSK_ESN);
    try {
        factory->getCryptoContext(ctx, data);
        ADD_FAILURE() << "Should have thrown.";
    } catch (const MslEntityAuthException& e) {
        EXPECT_EQ(MslError::ENTITY_REVOKED, e.getError());
    }
}

}}} // namespace netflix::msl::util
